File: 03-longest-substring-without-repeating/solution.cpp
// Longest Substring Without Repeating Characters
// Time: O(n), Space: O(min(n, m)) where m is charset size

#include <string>
#include <unordered_map>
#include <iostream>
#include <algorithm>

using namespace std;

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> charIndex;
        int maxLen = 0;
        int left = 0;
        
        for (int right = 0; right < s.length(); right++) {
            char c = s[right];
            
            if (charIndex.find(c) != charIndex.end() && charIndex[c] >= left) {
                left = charIndex[c] + 1;
            }
            
            charIndex[c] = right;
            maxLen = max(maxLen, right - left + 1);
        }
        
        return maxLen;
    }
};

// Test
int main() {
    Solution sol;
    cout << sol.lengthOfLongestSubstring("abcabcbb") << endl; // Output: 3
    cout << sol.lengthOfLongestSubstring("bbbbb") << endl;    // Output: 1
    cout << sol.lengthOfLongestSubstring("pwwkew") << endl;   // Output: 3
    return 0;
}

File: 03-longest-substring-without-repeating/solution.go
// Longest Substring Without Repeating Characters
// Time: O(n), Space: O(min(n, m)) where m is charset size

package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
    charIndex := make(map[rune]int)
    maxLen := 0
    left := 0
    
    for right, char := range s {
        if idx, found := charIndex[char]; found && idx >= left {
            left = idx + 1
        }
        
        charIndex[char] = right
        if right - left + 1 > maxLen {
            maxLen = right - left + 1
        }
    }
    
    return maxLen
}

// Test
func main() {
    fmt.Println(lengthOfLongestSubstring("abcabcbb")) // Output: 3
    fmt.Println(lengthOfLongestSubstring("bbbbb"))    // Output: 1
    fmt.Println(lengthOfLongestSubstring("pwwkew"))   // Output: 3
}
